smac/intensification/hyperband.py

```diff
@@ -0,0 +1,167 @@
+import logging
+import typing
+
+import numpy as np
+
+from smac.intensification.successive_halving import SuccessiveHalving
+from smac.stats.stats import Stats
+from smac.utils.constants import MAXINT
+from smac.configspace import Configuration
+from smac.runhistory.runhistory import RunHistory
+from smac.tae.execute_ta_run import ExecuteTARun
+from smac.utils.io.traj_logging import TrajLogger
+
+__author__ = "Ashwin Raaghav Narayanan"
+__copyright__ = "Copyright 2018, ML4AAD"
+__license__ = "3-clause BSD"
+
+
+class Hyperband(SuccessiveHalving):
+    """ Races multiple challengers against an incumbent using Hyperband method
+
+    Implementation from "BOHB: Robust and Efficient Hyperparameter Optimization at Scale" (Falkner et al. 2018)
+
+    Parameters
+    ----------
+    tae_runner : tae.executre_ta_run_*.ExecuteTARun* Object
+        target algorithm run executor
+    stats: Stats
+        stats object
+    traj_logger: TrajLogger
+        TrajLogger object to log all new incumbents
+    rng : np.random.RandomState
+    instances : typing.List[str]
+        list of all instance ids
+    instance_specifics : typing.Mapping[str,np.ndarray]
+        mapping from instance name to instance specific string
+    cutoff : int
+        runtime cutoff of TA runs
+    deterministic : bool
+        whether the TA is deterministic or not
+    min_budget : float
+        minimum budget allowed for 1 run of successive halving
+    max_budget : float
+        maximum budget allowed for 1 run of successive halving
+    eta : float
+        'halving' factor after each iteration in a successive halving run. Defaults to 3
+    run_obj_time : bool
+        whether the run objective is runtime or not (if true, apply adaptive capping)
+    n_seeds : int
+        Number of seeds to use, if TA is not deterministic. Defaults to None, i.e., seed is set as 0
+    instance_order : str
+        how to order instances. Can be set to:
+        None - use as is given by the user
+        shuffle_once - shuffle once and use across all SH run (default)
+        shuffle - shuffle before every SH run
+    adaptive_capping_slackfactor : float
+        slack factor of adpative capping (factor * adpative cutoff)
+    """
+
+    def __init__(self, tae_runner: ExecuteTARun,
+                 stats: Stats,
+                 traj_logger: TrajLogger,
+                 rng: np.random.RandomState,
+                 instances: typing.List[str],
+                 instance_specifics: typing.Mapping[str, np.ndarray] = None,
+                 cutoff: int = None,
```

Comment:
Please update all the signatures here to reflect that they allow `None` by annotating them with `typing.Optional[float]` etc...